File: CBLClient/Apps/JavaRequestHandlers/src/main/java/com/couchbase/mobiletestkit/javacommon/RequestHandler/VectorSearchRequestHandler.java
package com.couchbase.mobiletestkit.javacommon.RequestHandler;

import java.util.List;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import com.couchbase.mobiletestkit.javacommon.*;
import com.couchbase.lite.*;
import com.couchbase.lite.internal.utils.FileUtils;
import com.couchbase.mobiletestkit.javacommon.util.Log;

public class VectorSearchRequestHandler {
    private static  Map<String, Array> wordMap;
    private static final Boolean useInMemoryDb = true;
    private static final String inMemoryDbName = "InMemoryDb";
    
    static Map<String, Array> getWordVectMap() {
        try {
            DatabaseRequestHandler dbHandler = new DatabaseRequestHandler();
            Database db = preparePredefinedDatabase(inMemoryDbName);

            String sql1 = String.format("select word, vector from auxiliaryWords");
            Query query1 = db.createQuery(sql1);
            ResultSet rs1 = query1.execute();
            String sql2 = String.format("select word, vector from searchTerms");
            Query query2 = db.createQuery(sql2);
            ResultSet rs2 = query2.execute();

            Map<String, Array> words = new HashMap<String, Array>();
            List<Result> rl = rs1.allResults();
            List<Result> rl2 = rs2.allResults();
            rl.addAll(rl2);

            for (Result r : rl) {
                String word = r.getString("word");
                Array vector = r.getArray("vector");
                words.put(word, vector);
            }
            return words;

        } catch (Exception e) {
            System.err.println(e + "retrieving vector could not be done - getWordVector query returned no results");
            return null;
        }
    }

    public String createIndex(Args args) throws CouchbaseLiteException, Exception {
        Database database = args.get("database");
        String scopeName = args.get("scopeName") != null ? args.get("scopeName") : "_default";
        String collectionName = args.get("collectionName") != null ? args.get("collectionName") : "_default";

        Collection collection = database.getCollection(collectionName, scopeName);
        if (collection == null) {
            throw new Exception("Could not open specified collection");
        }

        String indexName = args.get("indexName");

        String expression = args.get("expression");

        int dimensions = args.get("dimensions");
        int centroids = args.get("centroids");
        Boolean isLazy = args.get("isLazy");

        VectorEncoding.ScalarQuantizerType scalarEncoding = args.get("scalarEncoding");

        Integer subquantizers = args.get("subquantizers");
        Integer bits = args.get("bits");

        String metric = args.get("metric");

        Integer minTrainingSize = args.get("minTrainingSize");
        Integer maxTrainingSize = args.get("maxTrainingSize");

        if (scalarEncoding != null && (bits != null || subquantizers != null)) {
            throw new Exception(
                    "Cannot have scalar quantization and arguments for product quantization at the same time");
        }

        if ((bits != null && subquantizers == null) || (bits == null && subquantizers != null)) {
            throw new Exception("Product quantization requires both bits and subquantizers set");
        }

        VectorIndexConfiguration config = new VectorIndexConfiguration(expression, dimensions, centroids);
        if (scalarEncoding != null) {
            config.setEncoding(VectorEncoding.scalarQuantizer(scalarEncoding));
        }
        if (bits != null) {
            config.setEncoding(VectorEncoding.productQuantizer(subquantizers, bits));
        }
        if (metric != null) {
            switch (metric) {
                case "euclidean":
                    config.setMetric(VectorIndexConfiguration.DistanceMetric.EUCLIDEAN);
                    break;
                case "cosine":
                    config.setMetric(VectorIndexConfiguration.DistanceMetric.COSINE);
                    break;
                default:
                    throw new Error("Invalid distance metric");
            }
        }

        if (minTrainingSize != null) {
            config.setMinTrainingSize(minTrainingSize);
        }

        if (maxTrainingSize != null) {
            config.setMaxTrainingSize(maxTrainingSize);
        }
    /*    if (isLazy != null) {
            config.isLazy = isLazy;
        }*/
        try {
            collection.createIndex(indexName, config);
            return String.format("Created index with name %s on collection %s", indexName, collectionName);
        } catch (Exception e) {
            return "Could not create index: " + e;
        }
    }

   /*  public String updateIndex(Args args) {
        String collection = args.get("collection");
        String indexName = args.get("indexName");
        QueryIndex config = new QueryIndex(collection, indexName);
        return "Temp dummy return";
    }*/

    public String registerModel(Args args) {
        String key = args.get("key");
        String name = args.get("name");
        vectorModel model = new vectorModel(key, inMemoryDbName);
        Database.prediction.registerModel(name, model);
        return "Registered model with name " + name;
    }

    public List<Object> query(Args args) throws CouchbaseLiteException, IOException {
        String term = args.get("term");

        Args embeddingArgs = new Args();
        embeddingArgs.put("input", term);
        Object embeddedTerm = this.getEmbedding(embeddingArgs);

        String sql = args.get("sql");

        Database db = args.get("database");

        Parameters params = new Parameters();
        params.setValue("vector", embeddedTerm);
        Query query = db.createQuery(sql);
        query.setParameters(params);

        List<Object> resultArray = new ArrayList<>();
        try (ResultSet rs = query.execute()) {
            for (Result row : rs) {
                resultArray.add(row.toMap());
            }
        }

        return resultArray;
    }

    public Object getEmbedding(Args args) throws CouchbaseLiteException, IOException {
        vectorModel model1 = new vectorModel("word", inMemoryDbName);
        MutableDictionary testDic = new MutableDictionary();
        String input = args.get("input");
        testDic.setValue("word", input);
        Dictionary value = model1.predict(testDic);
        return value.getValue("vector");
    }

    public Database loadDatabase(Args args) throws CouchbaseLiteException, IOException {
        if (useInMemoryDb) {
            wordMap = getWordVectMap();
            Database db = new Database(inMemoryDbName);
            return db;
        }
        else {
            Database db = preparePredefinedDatabase("dummtDBIgnoreIt");
            return db;
        }
    }

    private class vectorModel implements PredictiveModel {
        String key;
        String dbName;

        vectorModel(String key, String name) {
            try {
                this.key = key;
                this.dbName = name;
            } catch (Exception e) {
                System.err.println(e + "Error creating instance of vectorModel");
            }
        }

        @Override
        public Dictionary predict(Dictionary input) {
            String inputWord = input.getString(this.key);
            Object result = new ArrayList<>();
            result = wordMap.get(inputWord);
            MutableDictionary output = new MutableDictionary();
            output.setValue("vector", result);
            return output;
        }

    }

    private static Database preparePredefinedDatabase(String dbName) throws CouchbaseLiteException, IOException {
         // loads the given database vsTestDatabase
         DatabaseRequestHandler dbHandler = new DatabaseRequestHandler();
         Args newArgs = new Args();
         newArgs.put("dbPath", "vstestDatabase.cblite2");
         String dbPath = dbHandler.getVectorSearchDb(newArgs);
         newArgs.put("directory", "");
         if (RequestHandlerDispatcher.context.getPlatform().equals("android")) {
             newArgs.put("directory", new File(dbPath).getParent());
         }
         Database.exists("vstestDatabase.cblite2", new File(dbPath));
         DatabaseConfigurationRequestHandler configHandler = new DatabaseConfigurationRequestHandler();
         DatabaseConfiguration dbConfig = new DatabaseConfiguration();
         dbConfig = configHandler.configure(newArgs);
         newArgs.put("dbPath", dbPath);
         newArgs.put("dbName", dbName);
         newArgs.put("dbConfig", dbConfig);
         dbHandler.copy(newArgs);
         Database db1 = new Database(dbName);

         return db1;
    }
}

File: CBLClient/Apps/CBLTestServer-iOS/CBLTestServer-iOS/Server/VectorSearchRequestHandler.swift
//
//  VectorSearchRequestHandler.swift
//  CBLTestServer-iOS
//
//  Created by Monty Ali on 13/02/2024.
//  Copyright © 2024 Raghu Sarangapani. All rights reserved.
//

import Foundation
import CouchbaseLiteSwift
import CoreML
import Tokenizers
import Hub
#if COUCHBASE_ENTERPRISE
public class VectorSearchRequestHandler {
    public func handleRequest(method: String, args: Args) async throws -> Any? {
        switch method {

        // create index on collection
        case "vectorSearch_createIndex":
            guard let database: Database = args.get(name: "database") else { throw RequestHandlerError.InvalidArgument("Invalid database argument")}

            let scopeName: String = (args.get(name: "scopeName")) ?? "_default"
            let collectionName: String = (args.get(name: "collectionName")) ?? "_default"

            guard let collection: Collection = try database.collection(name: collectionName, scope: scopeName) else { throw RequestHandlerError.IOException("Could not open specified collection")}

            guard let indexName: String = args.get(name: "indexName") else { throw RequestHandlerError.InvalidArgument("Invalid index name")}

            guard let expression: String = args.get(name: "expression") else { throw RequestHandlerError.InvalidArgument("Invalid expression for vector index")}

            // IndexConfiguration type is UInt32 but Serializer does not handle UInt currently so use Int for now and assume always pass +ve numbers
            // For manual testing, note that serialization of ints uses format Ixx where xx is an int value
            guard let dimensions: Int = args.get(name: "dimensions") else { throw RequestHandlerError.InvalidArgument("Invalid dimensions argument for vector index")}

            // As above for dimensions, UInt32 vs Int
            guard let centroids: Int = args.get(name: "centroids") else { throw RequestHandlerError.InvalidArgument("Invalid centroids argument for vector index")}

            let scalarEncoding: ScalarQuantizerType? = args.get(name: "scalarEncoding")
            // UInt32/Int problem
            // Note that subquantizers needs to be a factor of dimensions, runtime error if not
            // and bits b needs 4 <= b <= 12, maybe worth adding input validation in future
            let subquantizers: Int? = args.get(name: "subquantizers")
            let bits: Int? = args.get(name: "bits")
            let metric: String? = args.get(name: "metric")
            // UInt32/Int problem
            let minTrainingSize: Int? = args.get(name: "minTrainingSize")
            let maxTrainingSize: Int? = args.get(name: "maxTrainingSize")

            if scalarEncoding != nil && (bits != nil || subquantizers != nil) {
                throw RequestHandlerError.InvalidArgument("Cannot have scalar quantization and arguments for product quantization at the same time")
            }

            if (bits != nil && subquantizers == nil) || (bits == nil && subquantizers != nil) {
                throw RequestHandlerError.InvalidArgument("Product quantization requires both bits and subquantizers set")
            }

            var config = VectorIndexConfiguration(expression: expression, dimensions: UInt32(dimensions), centroids: UInt32(centroids))
            if let scalarEncoding {
                config.encoding = VectorEncoding.scalarQuantizer(type: scalarEncoding)
            }
            if let bits {
                config.encoding = VectorEncoding.productQuantizer(subquantizers: UInt32(subquantizers!), bits: UInt32(bits))
            }
            if let metric {
                switch metric {
                case "euclidean":
                    config.metric = DistanceMetric.euclidean
                case "cosine":
                    config.metric = DistanceMetric.cosine
                default:
                    throw RequestHandlerError.InvalidArgument("Invalid distance metric")
                }
            }
            if let minTrainingSize {
                config.minTrainingSize = UInt32(minTrainingSize)
            }
            if let maxTrainingSize {
                config.maxTrainingSize = UInt32(maxTrainingSize)
            }
            do {
                try collection.createIndex(withName: indexName, config: config)
                return "Created index with name \(indexName) on collection \(collectionName)"
            } catch {
                return "Could not create index: \(error)"
            }

        // returns the embedding for input string
        case "vectorSearch_getEmbedding":
            let model = vectorModel(key: "test")
            let testDic = MutableDictionaryObject()
            guard let input: String = args.get(name: "input") else { throw RequestHandlerError.InvalidArgument("Invalid input for prediction")}
            testDic.setValue(input, forKey: "test")
            let prediction = model.predict(input: testDic)
            let value = prediction?.array(forKey: "vector")
            if let value {
                return value.toArray()
            } else {
                return "Could not generate embedding"
            }

        // register model that creates embeddings on the field referred to by key
        case "vectorSearch_registerModel":
            guard let key: String = args.get(name: "key") else { throw RequestHandlerError.InvalidArgument("Invalid key")}
            guard let name: String = args.get(name: "name") else { throw RequestHandlerError.InvalidArgument("Invalid name")}
            let model = vectorModel(key: key)
            Database.prediction.registerModel(model, withName: name)
            return "Registered model with name \(name)"

        // this is a very bare bones sql++ query handler which expects the
        // user to pass the query as a string in the request body
        // and simply tries to create a query from that string
        // will fail for invalid query input
        // assumes that you already have created an index
        case "vectorSearch_query":
            // term is the search query that will be embedded and
            // queried against
            guard let term: String = args.get(name: "term") else { throw RequestHandlerError.InvalidArgument("Invalid search term")}

            // internal call to handler to get vector embedding for search term
            let embeddingArgs = Args()
            embeddingArgs.set(value: term, forName: "input")
            let embeddedTerm = try await self.handleRequest(method: "vectorSearch_getEmbedding", args: embeddingArgs)

            // the sql query, paramaterised by $vector which will be
            // the embedded term
            guard let sql: String = args.get(name: "sql") else { throw RequestHandlerError.InvalidArgument("Invalid sql string")}

            // database to execute query against
            // may change to handle collections
            guard let db: Database = args.get(name: "database") else { throw RequestHandlerError.InvalidArgument("Invalid database")}

            let params = Parameters()
            params.setValue(embeddedTerm, forName: "vector")
            let query = try db.createQuery(sql)
            query.parameters = params
            guard let queryResults = try? query.execute() else { throw RequestHandlerError.VectorPredictionError("Error executing SQL++ query")}

            var results: [[String:Any]] = []
            for result in queryResults {
                results.append(result.toDictionary())
            }

            return results

        // load pre generated vsTestDatabase
        case "vectorSearch_loadDatabase":
            let dbHandler = DatabaseRequestHandler()
            let preBuiltArgs = Args()
            preBuiltArgs.set(value: "Databases/vsTestDatabase.cblite2", forName: "dbPath")
            let dbPath = try dbHandler.handleRequest(method: "database_getPreBuiltDb", args: preBuiltArgs)
            preBuiltArgs.set(value: dbPath!, forName: "dbPath")
            preBuiltArgs.set(value: "vsTestDatabase", forName: "dbName")
            _ = try dbHandler.handleRequest(method: "database_copy", args: preBuiltArgs)
            let db: Database = try Database(name: "vsTestDatabase")
            return db
        // the handlers below are for manual testing purposes and do not
        // need to be replicated in the other test servers

        // get tokenized input
        case "vectorSearch_testTokenizer":
            guard let input: String = args.get(name: "input") else { throw RequestHandlerError.InvalidArgument("Invalid input")}
            return try tokenizeInput(input: input)

        // tokenize input then decode back to string, including padding
        case "vectorSearch_testDecode":
            guard let input: String = args.get(name: "input") else { throw RequestHandlerError.InvalidArgument("Invalid input")}
            let tokens = try tokenizeInput(input: input)
            let decoded = try decodeTokenIds(encoded: tokens)
            return ["tokens": tokens, "decoded": decoded]


        default:
            throw RequestHandlerError.MethodNotFound(method)
        }
    }

}

@available(iOS 16.0, *)
public class vectorModel: PredictiveModel {
    // key is the field name in the doc
    // to create embeddings on
    let key: String
    let model = try! float32_model()

    public func predict(input: DictionaryObject) -> DictionaryObject? {
        guard let text = input.string(forKey: self.key) else { return nil }
        let result = MutableDictionaryObject()
        do {
            let tokens = try tokenizeInput(input: text)
            let modelTokens = convertModelInputs(feature: tokens)
            let attentionMask = generateAttentionMask(tokenIdList: tokens)
            let modelMask = convertModelInputs(feature: attentionMask)
            let modelInput = float32_modelInput(input_ids: modelTokens, attention_mask: modelMask)
            let embedding = try model.prediction(input: modelInput)
            let vectorArray = castToDoubleArray(embedding.pooler_output)
            result.setValue(vectorArray, forKey: "vector")
            return result
        } catch {
            return nil
        }
    }

    init(key: String) {
        self.key = key
    }
}

// Reads a config from Files/ directory
func readConfig(name: String) throws -> Config? {
    if let url = Bundle.main.url(forResource: "Files/\(name)", withExtension: "json") {
        do {
            let data = try Data(contentsOf: url, options: .mappedIfSafe)
            let jsonResult = try JSONSerialization.jsonObject(with: data, options: .mutableLeaves)
            if let jsonDict = jsonResult as? [NSString: Any] {
                return Config(jsonDict)
            }
        } catch {
            throw RequestHandlerError.IOException("Error retrieving json config: \(error)")
        }
    }
    return nil
}

// might need validation on input size, gte-small model only
// takes up to 128 tokens, so need to truncate otherwise
func tokenizeInput(input: String) throws -> [Int] {
    guard let tokenizerConfig = try readConfig(name: "tokenizer_config") else { throw RequestHandlerError.IOException("Could not read config") }
    guard let tokenizerData = try readConfig(name: "tokenizer") else { throw RequestHandlerError.IOException("Could not read config") }
    let tokenizer = try! AutoTokenizer.from(tokenizerConfig: tokenizerConfig, tokenizerData: tokenizerData)
    let tokenized = tokenizer.encode(text: input)
    return padTokenizedInput(tokenIdList: tokenized)
}

// refactor tokenizer stuff
func decodeTokenIds(encoded: [Int]) throws -> String {
    guard let tokenizerConfig = try readConfig(name: "tokenizer_config") else { throw RequestHandlerError.IOException("Could not read config") }
    guard let tokenizerData = try readConfig(name: "tokenizer") else { throw RequestHandlerError.IOException("Could not read config") }
    let tokenizer = try! AutoTokenizer.from(tokenizerConfig: tokenizerConfig, tokenizerData: tokenizerData)
    let decoded = tokenizer.decode(tokens: encoded)
    return decoded
}

func padTokenizedInput(tokenIdList: [Int]) -> [Int] {
    // gte-small constants
    let modelInputLength = 128
    let padTokenId = 0
    let inputNumTokens = tokenIdList.count
    var paddedTokenList = [Int]()

    paddedTokenList += tokenIdList
    if inputNumTokens < modelInputLength {
        var numTokensToAdd = modelInputLength - inputNumTokens
        while numTokensToAdd > 0 {
            paddedTokenList.append(padTokenId)
            numTokensToAdd -= 1
        }
    }
    return paddedTokenList
}

func generateAttentionMask(tokenIdList: [Int]) -> [Int] {
    var mask = [Int]()
    for i in tokenIdList {
        if i == 0 {
            mask.append(0)
        } else {
            mask.append(1)
        }
    }
    return mask
}

// convert input tokens and attention mask to the correct shape
func convertModelInputs(feature: [Int]) -> MLMultiArray {
    let attentionMaskMultiArray = try? MLMultiArray(shape: [1, NSNumber(value: feature.count)], dataType: .int32)
    for i in 0..<feature.count {
        attentionMaskMultiArray?[i] = NSNumber(value: feature[i])
    }
    return attentionMaskMultiArray!
}

// https://stackoverflow.com/q/62887013
func castToDoubleArray(_ o: MLMultiArray) -> [Double] {
    var result: [Double] = Array(repeating: 0.0, count: o.count)
    for i in 0 ..< o.count {
        result[i] = o[i].doubleValue
    }
    return result
}

private extension VectorSearchRequestHandler {
    // internal function for copying documents to the test database
    func copyWordDocument(copyFrom: Collection, copyTo: Collection, docIds: [String]) throws {
        let copyToCollectionName = copyTo.name
        for docId in docIds {
            guard let doc = try copyFrom.document(id: docId) else { throw RequestHandlerError.IOException("Could not get document \(docId)")}
            var docData = doc.toDictionary()

            if copyToCollectionName == "indexVectors" {
                docData.removeValue(forKey: "vector")
            }

            let newDoc = MutableDocument(id: docId, data: docData)
            try copyTo.save(document: newDoc)
        }
    }
}
#endif